### PR TITLE
Added requirement for 24h break every week

### PR DESCRIPTION
**Pull request by Quazyrog**
- Added requirement that on every full week every nurse has at least once 24h of continuous break
- Working weeks distribution included in the objective function
- New small input to test the 24h break requirement
- Includes changed names in model

### CODE REVIEW

File: model/main.mod [hunk 1]
```diff
@@ -25,11 +27,22 @@ set sUnpreferredCompanions within {sNurses, sNurses};
 set sPreferredSlots within {sNurses, sDays, sShifts};
 set sUnpreferredSlots within {sNurses, sDays, sShifts};
 
+# Binary indicator whether S consecutive shifts starting from given shift 
+# on given day are free for given nurse. 
+# This assumes that S shifts cover entire day, however length can be arbitrary.
+var vRest24hIndicator{sNurses, sWeeks, sWeekdays, sShifts}, binary;
+# Maximal allowed value of sum of vRest24hIndicator[n, w, *, *]:
+# there are S in each of 7 days; one oth them must be free
+param WORKED_24H_LIMIT := 6 * pNumberOfShifts;
```
Comment (czakis) [inline model/main.mod]: Does not follow `p<name>` naming convention + I think something along the lines of `pMaxDaysWithout24Break` would be more informative

File: model/main.mod [hunk 2]
```diff
@@ -62,9 +76,9 @@ subject to cWorkHoursLimit{n in sNurses}:
 subject to cDailyShiftLimit{nurse in sNurses, day in sDays}:
     sum{shift in sShifts} vSchedule[nurse, day, shift] <= 1;
     
-# respect night shifts per week limit
+# max 6 night shifts per week
```
Comment (czakis) [inline model/main.mod]: I'd revert to "respect night shifts per week limit"

File: model/main.tex [hunk 3]
```diff
@@ -37,7 +42,9 @@
     -&\lambda_{UC}\cdot\sum_{(n, n')\in UC}\sum_{d,s} vInteraction_{n,n',d,s} \\ 
     +&\lambda_{PS}\cdot\sum_{(n, d, s)\in PS} vSchedule_{n,d,s} \\ 
     -&\lambda_{US}\cdot\sum_{(n, d, s)\in US} vSchedule_{n,d,s} \\
-    -&(\alpha_{max} - \alpha_{min})%\quad \text{(tight alpha gap iff work distributed proportionately to wokrhours specified)}
+    -&(\alpha_{max} - \alpha_{min})\\%\quad \text{(tight alpha gap iff work distributed proportionately to wokrhours specified)}
```
Comment (czakis) [inline model/main.tex]: This line should be deleted

File: model/main.tex [hunk 4]
```diff
@@ -37,7 +42,9 @@
...
+     &\lambda_\text{WHS} (\alpha_{max} - \alpha_{min})\quad \text{(prefer equal work to max work hours ratio)} \\
+     &\lambda_\text{W} (\varMaxWeekendsWorked - \varMinWeekendsWorked)\quad \text{(prefer equal busy weekends distribution)}
```
Comment (czakis) [inline model/main.tex]: These weights should be with a minus

File: model/main.tex [hunk 5]
```diff
@@ -50,9 +57,13 @@
...
-&\forall_{n, w}\ vWeekend_{n, w} \leq \sum_s vSchedule_{n, 7(w-1)+6, s} + vSchedule_{n, 7(w-1)+7, s} \quad \text{(weekends 2)} \\
-&\forall_{n}\ vAlphaMin \leq \frac{pShiftLength\cdot\sum_{d,s}vSchedule_{n,d,s}}{pWorkhoursLimit_n} \leq vAlphaMax\quad \text{(alphas with bounds)}
+&\forall_{n}\ vAlphaMin \leq \frac{pShiftLength\cdot\sum_{d,s}vSchedule_{n,d,s}}{pWorkhoursLimit_n} \leq vAlphaMax\quad \text{(alphas with bounds)} \\
+&\forall_{n, w}\ vWeekend_{n, w} \geq \frac{1}{S}\sum_s (vSchedule_{n, 7w+6, s} + vSchedule_{n, 7w+7, s}) \quad \text{(weekends 1)} \\
+&\forall_{n, w}\ vWeekend_{n, w} \leq \sum_s (vSchedule_{n, 7w+6, s} + vSchedule_{n, 7w+7, s}) \quad \text{(weekends 2)} \\
+&\forall_{n}\ \varMinWeekendsWorked \le \sum_n \varWeekendWorkedIndicator{n}{s} \le \varMaxWeekendsWorked \quad \text{(min/max weekends worked are computed)} \\
+&\forall_{n,w,\delta,s}\ \varDayLengthBreakIndicator{n}{w}{\delta}{s} \ge \frac{\sum_{0 \le i \le pNumberOfShifts - 1} vSchedule_{n, d', s'}}{pNumberOfShifts}, \\
+    &\qquad\text{where } d' := 1+7w+\delta+\left\lfloor\frac{s+i}{pNumberOfShifts}\right\rfloor, s' := 1 + (s - 1 + i) \operatorname{mod} pNumberOfShifts \\
```
Comment (czakis) [inline model/main.tex]: It is not clear whether it is (1 + x) mod y or 1 + (x mod y)

File: model/main.tex [hunk 6]
```diff
@@ -6,14 +6,19 @@
 \usepackage[margin=2cm]{geometry}
 
 \newcommand{\N}{\mathbb{N}}
+\newcommand{\varWeekendWorkedIndicator}[2]{\texttt{\small vWeekend}_{#1, #2}}
+\newcommand{\varMinWeekendsWorked}{\texttt{\small vMinWeekendsWorked}}
+\newcommand{\varMaxWeekendsWorked}{\texttt{\small vMaxWeekendsWorked}}
+
+\newcommand{\varDayLengthBreakIndicator}[4]{\texttt{\small vRest24hIndicator}_{#1, #2, #3, #4}}
```
Comment (czakis) [inline model/main.tex:13]: Variable `vRest24hIndicator` is not described in "Optimization variables" section

**Review by czakis — COMMENTED** [review]
Looks good to me, there are few bugs in `.tex' file and some possible improvements in `main.mod`

File: model/main.mod (hunk 1 above)
Comment (Quazyrog) [reply model/main.mod]: So I will rename it to `pMaxDaysWithout24BreakInWeek`.

File: model/main.mod (hunk 2 above)
Comment (Quazyrog) [reply model/main.mod]: Right

**Pull request merged**